File: getting_started/binary_search.py
def search(data, key):
    idx_of_start = 0
    idx_of_end = len(data) - 1

    while idx_of_start <= idx_of_end:
        idx_of_mid = (idx_of_start + idx_of_end) // 2
        if data[idx_of_mid] == key:
            return idx_of_mid
        elif data[idx_of_mid] < key:
            idx_of_start = idx_of_mid + 1
        else:
            idx_of_end = idx_of_mid - 1

    return -1


if __name__ == '__main__':
    data = [i * 2 for i in range(20)]
    
    print(data)

    print("input key > ", end='')
    key = int(input())

    result = search(data, key)
    if result != -1:
        print(f"result = {result}")
    else:
        print(f"{key} is not found.")

File: getting_started/binary_search.go
//go:build ignore

package main

import (
	"fmt"
)

func search(data []int, key int) int {
	idxOfStart := 0
	idxOfEnd := len(data) - 1

	for idxOfStart <= idxOfEnd {
		idxOfMid := (idxOfStart + idxOfEnd) / 2
		if data[idxOfMid] == key {
			return idxOfMid
		} else if data[idxOfMid] < key {
			idxOfStart = idxOfMid + 1
		} else {
			idxOfEnd = idxOfMid - 1
		}
	}

	return -1
}

func main() {
	data := make([]int, 20)
	for i := 0; i < 20; i++ {
		data[i] = 2 * i
	}

	fmt.Println(data)

	fmt.Print("input search key > ")
	var key int
	fmt.Scanf("%d", &key)

	result := search(data, key)
	if result != -1 {
		fmt.Printf("result = %d\n", result)
	} else {
		fmt.Printf("%d is not found.", key)
	}
}
